### ISSUE TITLE
runtime error (7) : too many resources requested for launch at pytorch/aten/src/THC/THCTensorSort.cu:62

### ISSUE THREAD

**Issue by perry0418** [pytorch/pytorch#17144]
### THCudaCheck FAIL file=/home/nvidia/Documents/pytorch/aten/src/THC/THCTensorSort.cu line=62 error=7 : too many resources requested for launch
when I run the maskrcnn (project link: https://github.com/facebookresearch/maskrcnn-benchmark) inference demo on nvidia jetson TX2, and it occurs the runtime error (7). by the way, I also test the YOLO v3 network on TX2, which works well. 

* PyTorch : torch-1.1.0a0+7c66ad7
* install PyTorch : from source
* Python version: 3.5
* CUDA/cuDNN version: cuda9.0 and cudnn7.1.5 (original from TX2 jetpack 3.3)

### error log:
THCudaCheck FAIL file=/home/nvidia/Documents/pytorch/aten/src/THC/THCTensorSort.cu line=62 error=7 : too many resources requested for launch
Traceback (most recent call last):
  File "demo.py", line 88, in <module>
    predictions = coco_demo.run_on_opencv_image(image)
  File "/home/nvidia/Documents/maskrcnn-benchmark-master/demo/predictor.py", line 93, in run_on_opencv_image
    predictions = self.compute_prediction(image)
  File "/home/nvidia/Documents/maskrcnn-benchmark-master/demo/predictor.py", line 124, in compute_prediction
    predictions = self.model(image_list)
  File "/usr/local/lib/python3.5/dist-packages/torch/nn/modules/module.py", line 492, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/nvidia/Documents/maskrcnn-benchmark-master/maskrcnn_benchmark/modeling/detector/generalized_rcnn.py", line 50, in forward
    proposals, proposal_losses = self.rpn(images, features, targets)
  File "/usr/local/lib/python3.5/dist-packages/torch/nn/modules/module.py", line 492, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/nvidia/Documents/maskrcnn-benchmark-master/maskrcnn_benchmark/modeling/rpn/rpn.py", line 96, in forward
    return self._forward_test(anchors, objectness, rpn_box_regression)
  File "/home/nvidia/Documents/maskrcnn-benchmark-master/maskrcnn_benchmark/modeling/rpn/rpn.py", line 122, in _forward_test
    boxes = self.box_selector_test(anchors, objectness, rpn_box_regression)
  File "/usr/local/lib/python3.5/dist-packages/torch/nn/modules/module.py", line 492, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/nvidia/Documents/maskrcnn-benchmark-master/maskrcnn_benchmark/modeling/rpn/inference.py", line 138, in forward
    sampled_boxes.append(self.forward_for_single_feature_map(a, o, b))
  File "/home/nvidia/Documents/maskrcnn-benchmark-master/maskrcnn_benchmark/modeling/rpn/inference.py", line 93, in forward_for_single_feature_map
    objectness, topk_idx = objectness.topk(pre_nms_top_n, dim=1, sorted=True)
RuntimeError: cuda runtime error (7) : too many resources requested for launch at /home/nvidia/Documents/pytorch/aten/src/THC/THCTensorSort.cu:62

**Comment by fmassa** [comment]
cc @t-vi, who is porting topk to ATen. There is maybe one of the hard-coded parameters which doesn't fit a TX2?
Also, @ngimel do you happen to know a better set of parameters?

**Comment by t-vi** [comment]
The main thing I'm aware of is that the switchover from bitonic sort to thrust as a fallback is different for 64 bit dtypes. These have been lowered for sort but that has not been applied to topk's sorting.
The natural questions would be
- What is the tensor dtype and size and what are dim and k used ins topk?
- Does the non-sorting topk work for that?
- Does sort work for the non-sorted values?
(The last two also might offer a workaround util it is fixed in PyTorch).

**Comment by perry0418** [comment]
hi @t-vi，thanks for your advice , I test the program follow your suggestion. and post the log at below. I think there may have some parameter of topk need fix in Pytorch for TX2.
### 1. the tensor size and dtype are printed as below 
print the tensor size:
torch.Size([1,50250]) 
print the tensor type:
torch.cuda.FloatTensor
print pre_nms_top_n:
6000
objectness, topk_idx = objectness.topk(pre_nms_top_n, dim=1, sorted=True)

### 2. I change the sorted=False, it still dose not work.  the error log :
File "/home/nvidia/Documents/maskrcnn-benchmark-master/maskrcnn_benchmark/modeling/rpn/inference.py", line 93, in forward_for_single_feature_map
    objectness, topk_idx = objectness.topk(pre_nms_top_n, dim=1, sorted=False)
RuntimeError: cuda runtime error (7) : too many resources requested for launch at /home/nvidia/Documents/pytorch/aten/src/THC/generic/THCTensorTopK.cu:178

**Comment by t-vi** [comment]
Thanks for trying this out!
Then we need to set launch bounds on the topk kernels.
Given that my ATen move is a bit stuck, I'll send a targeted fix.

**Comment by t-vi** [comment]
@perry0418 Can you check if the [branch above](https://github.com/t-vi/pytorch/tree/topk_launch_bounds) fixes your problem, please? Then I'd make a PR for it.

**Comment by perry0418** [comment]
thx @t-vi,  I will check it soon , and give the feedback later.

**Comment by perry0418** [comment]
@t-vi 
thank you so much!
it works !

**Comment by t-vi** [comment]
Can you reopen the issue until so we can track fixing it, please?

**Comment by kealennieh** [comment]
sorry, I've met the same problem. My pytorch version is v1.0.0. Therefore, I replaced the two files mentioned above. However, I've got more errors. 
/pytorch/aten/src/THC/THCTensorTopK.cuh(364): error: expected a type specifier
/pytorch/aten/src/THC/THCTensorTopK.cuh(364): error: explicit type is missing ("int" assumed)
/pytorch/aten/src/THC/THCTensorTopK.cuh(364): error: expected a ";"
.....

any suggestions ? Many thanks.
